### PR TITLE
Fix params for saved card csv downloading

### LINKED ISSUE #3695: CSV Download for SQL Query Returns Default Value on Variable
### Bugs

If you're filing an issue about a bug please include as much information
as you can including the following.
- Chrome 53.0.2785.143
- OSX 10.10
- Postgres, MySQL
- Metabase version v0.20.2
- Metabase hosting environment: Ubuntu16.04
- Metabase internal database: Postgres 9.5

_Repeatable steps to reproduce the issue_
1. Create new card with SQL query with below sample from database **Sample Dataset**:
screen shot 2016-10-28 at 10 42 50 am
`select * from orders
where user_id={{user_id}}`
Then I set default value of user_id into 2.
2. Save the question, reopen it and change the value of variable (example: change the user_id into 3). Everything went well until this step:
screen shot 2016-10-28 at 10 45 58 am
3. Click download CSV, and when we open the CSV, it contains user_id 2 (default value of variable user_id). It doesn't respect variable value that we choose. Here is the result from CSV download:
screen shot 2016-10-28 at 10 49 16 am

Okay let's try another case not to set the default value like below card:
1. Create card with sql query and a required variable (without default value)
screen shot 2016-10-28 at 10 51 23 am
2. Reopen the card, set the variable value, and click download csv:
screen shot 2016-10-28 at 10 53 17 am
screen shot 2016-10-28 at 10 53 24 am
We cannot download the CSV  with error `'User' is a required param.`

---

### PR DESCRIPTION
**Pull request by tlrobinson**
Resolves #3695

### CODE REVIEW

**Comment by salsakran** [comment]
can we get a test for this?

File: src/metabase/api/card.clj [hunk 1]
```diff
@@ -273,9 +274,11 @@
   (run-query-for-card card-id parameters))
 
 (defendpoint POST "/:card-id/query/csv"
-  "Run the query associated with a Card, and return its results as CSV."
-  [card-id :as {{:keys [parameters]} :body}]
-  (dataset-api/as-csv (run-query-for-card card-id parameters)))
+  "Run the query associated with a Card, and return its results as CSV. Note that this expects the query as serialized JSON in the 'query' parameters."
+  [card-id query]
+  {query (s/maybe su/JSONString)}
+  (dataset-api/as-csv (run-query-for-card card-id (when query
```
Comment (camsaul) [inline src/metabase/api/card.clj]: You don't actually need to wrap it in the when, parse-string will return nil if it's input is nil

**Comment by tlrobinson** [comment]
@salsakran Are you looking for an integration test or backend test?

**Comment by salsakran** [comment]
either works for me. 

an integration test would be nice in that it also verifies that the front-end is calling the api sensibly.

**Comment by tlrobinson** [comment]
I looked into writing an integration test and I don't think there's a way to check that a file was downloaded using Selenium. If we weren't using Sauce (thus running the browser on a different machine than the tests), we could check the filesystem, but that would have it's own problems.

@camsaul I could use some help writing a test that hits the csv endpoints with parameters and making sure they're applied.

**Comment by camsaul** [comment]
I'll look into writing some new tests give me a bit I'll add them to this branch

**Comment by camsaul** [comment]
@tlrobinson why did you change the endpoint to take a whole `query` instead of just the params? That's misleading for other people who might try to use the API because the query input is ignored and isn't even validated or anything like that

**Comment by camsaul** [comment]
It's also confusing because the auto-generated documentation will say that `:query` is an optional param for the endpoint, instead of just `:parameters`

**Comment by tlrobinson** [comment]
Yeah I was trying to keep it consistent with the other CSV endpoint but I think you're right. I can change the frontend if you want to change the endpoint.

**Comment by camsaul** [comment]
@tlrobinson cool WFM. I'll change the backend stuff 👍

**Comment by camsaul** [comment]
ok @tlrobinson I switched the Card CSV & JSON endpoints to take `?parameters=` instead of `query` and added tests for them 👍

**Comment by tlrobinson** [comment]
I've already changed all the frontend code that this change will affect in #3779 so I'd rather just take care of this there and merge it all together. That PR is almost done anyway.

**Pull request closed without merging**